build-system/compile/closure-compile.js

```diff
@@ -15,119 +15,119 @@
+ * @param {string} err
  * @param {string} outputFilename
  * @param {?Object} options
- * @param {?Function} resolve
  */
-function handleCompilerError(err, outputFilename, options, resolve) {
-  logError(`${red('ERROR:')} Could not minify ${cyan(outputFilename)}`);
+function handleCompilerError(err, outputFilename, options) {
+  const message = `${red('ERROR:')} Could not minify ${cyan(outputFilename)}`;
+  logError(message, err);
   if (options && options.continueOnError) {
     options.errored = true;
-    if (resolve) {
-      resolve();
-    }
   } else {
-    emitError(err);
+    throw new Error(message);
   }
 }
 
 /**
  * Handles a closure error during type checking
  *
- * @param {Error} err
+ * @param {string} err
  */
 function handleTypeCheckError(err) {
-  logError(red('Type checking failed:'));
-  emitError(err);
+  const message = `${red('ERROR:')} Type checking failed`;
+  logError(message, err);
+  throw new Error(message);
 }
 
 /**
- * Emits an error to the caller
- *
- * @param {Error} err
+ * Prints an error message when compilation fails
+ * @param {string} message
+ * @param {string} err
  */
-function emitError(err) {
-  err.showStack = false;
-  new EventEmitter().emit('error', err);
+function logError(message, err) {
+  log(`${message}\n` + formatClosureCompilerError(err));
 }
 
 /**
- * Prints an error message when compilation fails
- * @param {string} message
+ * Extracts the actual error message from closure compiler's stdErr by removing
+ * the long preamble that ends with the full command line.
+ * @param {Array<string>} flags
+ * @param {string} stdErr
+ * @return {string}
  */
-function logError(message) {
-  log(`${message}\n` + formatClosureCompilerError(compilerErrors));
+function getErrorText(flags, stdErr) {
+  const lastFlag = flags.slice(-1).pop();
+  return stdErr.split(lastFlag).pop();
 }
 
 /**
- * Normalize the sourcemap file paths before pushing into Closure.
- * Closure don't follow Gulp's normal sourcemap "root" pattern. Gulp considers
- * all files to be relative to the CWD by default, meaning a file `src/foo.js`
- * with a sourcemap alongside points to `src/foo.js`. Closure considers each
- * file relative to the sourcemap. Since the sourcemap for `src/foo.js` "lives"
- * in `src/`, it ends up resolving to `src/src/foo.js`.
- *
- * @param {!NodeJS.WritableStream} closureStream
- * @return {!NodeJS.WritableStream}
+ * Initializes closure compiler with the given set of flags and ensures that the
+ * native compiler is used during compilation.
+ * @param {Array<string>} flags
+ * @return {!Object}
  */
-function makeSourcemapsRelative(closureStream) {
-  const relativeSourceMap = sourcemaps.mapSources((source, file) => {
-    const dir = path.dirname(file.sourceMap.file);
-    return path.relative(dir, source);
-  });
-
-  return pumpify.obj(relativeSourceMap, closureStream);
+function initializeClosure(flags) {
+  const closure = new compiler(flags);
+  closure.JAR_PATH = null;
```

Comment:
Having to reach into the implementation this way isn't great. Could we enhance the implementation in the other library we work in?

What happens on ARM64 machines (where there is no native binary support)?

Reply:
Entirely agree. (We don't run any of our tests on ARM today, but it is something we should support even if builds are slower. Will make a change upstream and update this code.)